### ISSUE TITLE
Implementing a version of the (Sparse) Categorical Cross Entropy Loss that makes use of Integer-Indexed predictions.

### ISSUE THREAD

**Issue by mayowaosibodu** [tensorflow/tensorflow#47655]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): Tensorflow 2.
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**

The Categorical Cross Entropy Loss makes use of both one-hot encoded predictions, and one-hot encoded true labels.
Eg, y_true = [ [0,0,1], [1,0,0] ] and y_pred = [ [0,0,1], [0,1,0] ].

The Sparse Categorical Cross Entropy Loss makes use of single integer true labels, but still requires an array of probabilities corresponding to each existing class, per label in the predictions.
Eg, y_true = [2, 0] and y_pred = [ [0.1, 0.2, 0.7], [0.2, 0.85, 0.05] ].

I feel it would be really useful to have a loss function which makes use of single integer-indexed categorical predictions as well as single integer-indexed categorical true labels.
Eg, y_true = [2, 0] and y_pred = [2, 1].

**Will this change the current api? How?**
Not really. It could possibly just require the definition of a new loss function.

**Who will benefit with this feature?**
For one, it would be possible to use Dense models to process two-dimensional data. This could be done by simply encoding one of the concerned dimensions as a series of integer-indexed categories. You could process text data with a Dense model by simply encoding the characters as integer-indexed categories (in contrast to one-hot encoding) and then work on minimizing the outputs of the proposed loss function.

Eg, "I am" would simply be [9,0,1,13] (with ' ' as 0). This is evidently amenable to processing by a Dense model.

**Any Other info.**

**Comment by rmothukuru** [comment]
@mayowaosibodu,
Do you mean to say that your request is to facilitate effective usage of `Dense Layers` for `Text Data`? If it is so, we have `Recurrent Neural Networks` (`RNNs`, `LSTMs`, `GRUs`, etc..) for the same. You can correct me if I'm missing something.

**Comment by mayowaosibodu** [comment]
@rmothukuru I see what you're saying. There exist models built specifically with sequence data in mind.

However I'm saying there appears to be away to make Dense-layer models capable of processing textual data, with very minimal changes to the current codebase.

I do not know of academic articles assessing the capabilities of Dense-layer models with regard to Text Data.
There do not even seem to be any accessible online, and I understand and somewhat expect that, given that sequence-specific models are now the standard for textual data.

But it just seems to me like a promising opportunity to explore. Dense-layer models are definitely free of all of the performance encumbrances that RNNs etc, experience. I've worked with Deep LSTM networks on pretty capable GPU-cloud servers before and I know they are computationally expensive (even more so than ConvNets built to process the same data). And so given how computationally frugal (relatively), Dense-layer models are, I personally think any opportunity to try them out on datasets that normally would be outside their purview is definitely one worth exploring.

Building off of your response, I realize there's definitely a more theoretical AI angle to this, given that it involves a technique that is pretty different from convention, but I feel like with regard to feature implementation, it's a very trivial issue. It's really just one function for the new loss.

--

By the way, there could possibly be other uses for this feature, but working with Text Data is the most forefront on my mind right now because it was within this context I thought of the idea in the first place. With regard to facilitated programming ease, I don't know if anyone else is complaining that one-hot encoding is bulky, so in that specific respect I'm not sure how much consensus exists there that such a feature is very needed.

**Comment by tilakrayal** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
